piff/gp_interp.py

```diff
@@ -13,113 +13,114 @@
+    :param n_neighbors:  Number of neighbors to used for interpolating the spatial average using
+                         a KNeighbors interpolation. Used only if average_fits is not None. [defaulf: 4]
+    :param logger:       A logger object for logging debug info. [default: None]
     """
-    def __init__(self, keys=('u','v'), kernel='RBF()', optimize=True, npca=0, normalize=True,
-                 logger=None):
-        from sklearn.gaussian_process import GaussianProcessRegressor
+    def __init__(self, keys=('u','v'), kernel='RBF(1)',
+                 optimizer='two-pcf', normalize=True, l0=3000.,
+                 white_noise=0., n_neighbors=4, average_fits=None,
+                 nbins=20, min_sep=None, max_sep=None,
+                 rows=None, logger=None):
 
         self.keys = keys
+        self.optimizer = optimizer
+        self.l0 = l0
+        self.n_neighbors = n_neighbors
+        self.average_fits = average_fits
+        self.nbins = nbins
+        self.min_sep = min_sep
+        self.max_sep = max_sep
+        self.normalize = normalize
+        self.white_noise = white_noise
         self.kernel = kernel
-        self.npca = npca
-        self.degenerate_points = False
+        self.rows = rows
 
         self.kwargs = {
             'keys': keys,
-            'optimize': optimize,
-            'npca': npca,
+            'optimizer': optimizer,
             'kernel': kernel
         }
-        optimizer = 'fmin_l_bfgs_b' if optimize else None
-        self.gp = GaussianProcessRegressor(self._eval_kernel(self.kernel), optimizer=optimizer,
-                                           normalize_y=normalize)
-
-    @staticmethod
-    def _eval_kernel(kernel):
-        # Some import trickery to get all subclasses of sklearn.gaussian_process.kernels.Kernel
-        # into the local namespace without doing "from sklearn.gaussian_process.kernels import *"
-        # and without importing them all manually.
-        def recurse_subclasses(cls):
-            out = []
-            for c in cls.__subclasses__():
-                out.append(c)
-                out.extend(recurse_subclasses(c))
-            return out
-        clses = recurse_subclasses(Kernel)
-        for cls in clses:
-            module = __import__(cls.__module__, globals(), locals(), cls)
-            execstr = "{0} = module.{0}".format(cls.__name__)
-            exec(execstr, globals(), locals())
-
-        from numpy import array
-
-        try:
-            k = eval(kernel)
-        except Exception as e:  # pragma: no cover
-            raise RuntimeError("Failed to evaluate kernel string {0!r}.  "
-                               "Original exception: {1}".format(kernel, e))
-        return k
-
-    def _fit(self, X, y, logger=None):
-        """Update the GaussianProcessRegressor with data
+
+        if isinstance(kernel,str):
+            self.kernel_template = [kernel]
+        else:
```

Comment:
This is not covered.  I think it can be removed.  The docs say that kernel should be a str, so as documented, only the True branch is allowed.

Reply:
This is the doc which is not up to date. I updated like that this is possible to have a kernel to be a list of kernel and like that to have a different type of kernel per PSF parameters. I added a test in the `test_gp_interp.py`